OMCompiler/Compiler/BackEnd/Tearing.mo

```diff
@@ -1648,100 +1671,163 @@ protected function minimalTearing
     end if;
+    qidx := qidx + 1;
   end for;
-  for i in eindex loop
-    eqArray[i] := true;
-  end for;
+  // print("Non-algorithm-output discrete Vars: " + stringDelimitList(List.map(unsolvedDiscreteVars,intString),",") + "\n");
 
-  (_,aMatrix,aMatrixT,_,_) := BackendDAEUtil.getAdjacencyMatrixScalar(isyst,BackendDAE.SOLVABLE(), SOME(ishared.functionTree), BackendDAEUtil.isInitializationDAE(ishared));
+  // Match the remaining discrete variables
+  if not listEmpty(unsolvedDiscreteVars) then
+    matchDiscreteVars(unsolvedDiscreteVars, adjEnhT, varArray, eqArray, nE, nV);
+    // make inner equations for the matched non-algorithm-output discrete vars.
+    (varArray, eqArray, innerEquations) := getTearingSetfromAssign(unsolvedDiscreteVars, nE, varArray, eqArray);
 
-  // Match discrete variables
-  if not listEmpty(discreteVars) then
-    matchDiscreteVars(discreteVars, isyst, ishared, aMatrix, aMatrixT, varArray, eqArray, nE, nV);
-    if debug then
-      print("\nDiscrete Vars:\n" + stringDelimitList(List.map(discreteVars,intString),",") + "\n\n");
-      print("\nnE Array:\n" + stringDelimitList(List.map(arrayList(nE),intString),",") + "\n\n");
-      print("\nnV Array:\n" + stringDelimitList(List.map(arrayList(nV),intString),",") + "\n\n");
-    end if;
-    //Tearing of the assign
-    (varArray, eqArray, helpInnerEquations) := getTearingSetfromAssign(discreteVars, nE, varArray, eqArray, helpInnerEquations);
-  end if;
+    for iq in innerEquations loop
+      innerEquationsLocalIndex := iq::innerEquationsLocalIndex;
+    end for;
 
-  if debug then
-    print("nE Array:\n" + stringDelimitList(List.map(arrayList(eqArray),boolString),",") + "\n\n");
-    print("nV Array:\n" + stringDelimitList(List.map(arrayList(varArray),boolString),",") + "\n\n");
-    print("Discrete Vars:\n" + stringDelimitList(List.map(discreteVars,intString),",") + "\n\n");
   end if;
 
-  // Tear discrete variables
-  for i in 1:size loop
+  // Mark all other equations as residual and all other vars
+  // as tearing vars.
+  // This can be improved a bit to be clearer.
+  for i in 1:listLength(eindex) loop
     if eqArray[i] then
       residualequations := i::residualequations;
     end if;
 
+    // This ordering of iteration vars based on the ordering
+    // in the normal adjacency matrix seems to cause differences
+    // in sumulation time.
+    // What is odd is that the order in adj matrix is not really guided
+    // by anything. Somehow this seems to be faster for some models.
+    // However, we now use the enahnced matrix which can sometimes have
+    // the var entries for an equation slightly different order than the normal
+    // adj matix.
+    // Simulation should not be affected by what order we have here but it is.
+    // There is something in later phases that operates according to this order
+    // But it should not. This is just random order.
+    /*
```

Comment:
I just had a discussion with @lochel about this. And his suggestion is that it can vary from model to model. What bothered me was the amount of variation seems to be so big. 

It seems there is something we have to further investigate on how the Jacobian handling creates the sparsity and color structure for the systems. It is one of the big differences I saw in the generated code.

The good news is that if you use **kinsol** as your solver then the performance difference disappears completely for all orderings.

It is a bit involved to explain here but I will try. So we are considering two ways of ordering for the iteration vars in this seemingly rudimentary integer list. Say we have three residual equations, 1,2 and 3. Eq 1 uses vars 2 and 3. Eq 2 uses vars 1 and 2. 
1 -> {2,3}
2 -> {1,2}
3 -> {1,2,3}

The original ordering we were creating for the iteration vars was (let's call this with LOCALITY)
2,3,1

because we go through the residual equations in order and add the variables they used. I thought this was ineffective so I changed it to just iterate over the variables and add the as iteration if they are not matched. So we get the order (let's call this INCREASING ORDER)
1,2,3

What is surprising is that if the matrix looked like this (let's say if we use the enhanced matrix)
1 -> {3,2}**
2 -> {1,2}
3 -> {1,2,3}

The LOCALITY ordering will be 
3,2,1

and this seems to have an effect on simulation performance. 


------------------------------------------------------------------
Here is a full example for on NLS of a [PowerGrids Model](https://trac.openmodelica.org/OpenModelica/ticket/5797)
------------------------------------------------------------
Non-Enhanced LOCALITY ORDERING
* tVars: 42,38,41,20,40,21,39,37,24,22,19,36,35,34,33,30,2,32,17,31,4,18,29,28,27,26,25,23,1,16,3,15,14,6,5

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim time - 1.954s
sim-kinsol - 0.749s**

Enhanced LOCALITY ORDERING
* tVars: 38,42,41,20,40,21,24,39,22,37,19,36,35,34,30,33,2,32,17,31,4,18,29,28,27,26,25,23,1,16,3,15,14,6,5

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim - 7.297s
sim-kinsol - 0.733s**

Enhanced INCREASING ORDER
* tVars: 1,2,3,4,5,6,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42

* resEq: 36,35,34,33,32,31,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1

**sim - 14.834s
sim-kinsol - 0.746s**


As you can see there is quite a big differnce in simulation time. What is even more intersting is that oredring of variables matters even for one single equation. E.g.  Residual equation 36 uses vars 42 and 38. But the normal and enahanced adj matrices order them differently as you said, {42,38} and {38,42} respectively. This changes in ordering take the simulation time from 1.9 sec to 7.297 secs.

I do not know enough to fully interpret what is happening but I though it should be at least highlighted.

Reply:
For sure, this is quite frightening... I guess these are bad quirks of dense solvers. Maybe we should talk about this in more detail when we have some time!

Reply:
In the case of this model, there is clearly no question that a sparse solver should be used, and there is absolutly no point in using a dense solver and trying to optimize it. Would you care about optimizing an oxen-powered carriage when you have a diesel truck at hand? 😄 

In fact, not only a dense solver should not be used, but we shouldn't even use odeMode, and rather use daeMode in this case, because only daeMode keeps the structure of the model sparse until the end.

What we do need it to set the defaults for the selection solvers and solver mode (ode vs. dae) so that the right one is selected automatically. First of all, if one really wanted to use odeMode, a sparse nonlinear solver should be used (currently we only have kinsol, which is ok). I guess the current default limits of size and density for this selection are too conservative in favour of dense solvers. 

Second, we should probably suggest to try daeMode instead. Dymola already does it. As soon as daeMode is good enough with events, see [#5860](https://trac.openmodelica.org/OpenModelica/ticket/5860)

Reply:
@casella I agree completely on the issue of choosing the proper solver. 

The reason I added those comments to the code was that we need to figure out (just in case) why the order of variable detection in a single equation has that much performance effect for dense solvers. This issue will be there for any tearing method, not just minimalTearing. If we change up how we traverse a simple expression `a+b` vs` b+a` there is a significant difference in performance. The dense solver seems to be significantly affected by this order. I assume that some other heuristic that deals with non-linear systems after tearing needs some revising. Because it seems to be noticeably inconsistent for very small differences.

Reply:
> 
> 
> @casella I agree completely on the issue of choosing the proper solver.
> 
> The reason I added those comments to the code was that we need to figure out (just in case) why the order of variable detection in a single equation has that much performance effect for dense solvers. This issue will be there for any tearing method, not just minimalTearing. If we change up how we traverse a simple expression `a+b` vs` b+a` there is a significant difference in performance. The dense solver seems to be significantly affected by this order. I assume that some other heuristic that deals with non-linear systems after tearing needs some revising. Because it seems to be noticeably inconsistent for very small differences.

I agree. However, I'm not sure I fully understand what is the difference in the three cases:
- are the tearing and torn variables the same?
- is the coloring of the Jacobian the same?
- is the variability in the ordering of torn equations made possible by the fact that only a partial ordering is enforced by the dependency graph?

Reply:
1. The torn/iteration variables chosen are the same.
2. The coloring of the jacobian seems to be different. At least from what I can see in the generated code.
3. Yes? If I interpreted this question correctly. The end result in any case is that the ordering in the list of (integer indices) of variables chosen as iteration variables affects the simulation performance considerably.

If you change the ordering, of course, you expect some change. I understand that but my assumption was that any processing that depends on such ordering (e.g. sparsity) should be improved to decide the detection and ordering itself. Because at tearing time, we can't really make an informed decision about this. a+b+c might be collected as {1,2,3} or {2,3,1} depending on the traversal functions we use (which are different for the two types of matrices we have). And this is what caused the difference in simulation time. 

Anyway this is just one model. And it might very well be something we can not avoid or is the expected behavior.

Reply:
Wait, I find it a bit odd that the Jacobian colouring is influenced by the ordering of the torn equations and variables.

I lack some basic theoretical understanding here, please help me with that. Should the colouring of the Jacobian be unique, or can there be legitimate different choices for that? In the latter case, I guess we should consider heuristics to make sure that as many colours as possible are generated. Was that ever taken into consideration?

Reply:
> I lack some basic theoretical understanding here
Imagine me then :). 

> please help me with that. Should the colouring of the Jacobian be unique, or can there be legitimate different choices for that?
I have no idea really.  i do not know much how the sparsity pattern or coloring is actually used. I was hoping you would know what was going on or should be expected actually.


> Wait, I find it a bit odd that the Jacobian colouring is influenced by the ordering of the torn equations and variables.
Yup. That is where I was curious as well. To make it worse, it is not just the ordering of the equations, but the ordering of the detection of used variables for each equation (which affects the final order)

> I guess we should consider heuristics to make sure that as many colours as possible are generated. Was that ever taken into consideration?
As far as I can see the "number of colors"? was different for the two orderings.

Reply:
@AnHeuermann, @kabdelhak , I guess we need your comment here 😃

Reply:
Taken from [What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf):

11.3. Bounds for the Chromatic Number. Since computing the chromatic number χ(G) of a general graph G is NP-hard, we consider bounding it with more tractable graph parameters.

The problem of (perfect) coloring is NP-Hard so it is computed by a heuristic which can change depending on the order of equations and variables.

Reply:
I don't know if there is a bug or if it is just that sensitive.

Reply:
On a slightly unrelated note 
[What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf): 
is actually written by my cousin. I guess I should read it from cover to cover.  😊😊

Reply:
> I don't know if there is a bug or if it is just that sensitive.

I'd say that the problem that raised @mahge attention is definitly overstretched to be handled by tearing and dense solvers. The question is whether or not we can get similar differences in performance (1:10) in problems where tearing and dense solvers are appropriate. Is there any means to compare the performance on the whole testsuite by running test jobs with some extra parameters?

Reply:
> 
> 
> On a slightly unrelated note
> [What Color Is Your Jacobian?](https://pdfs.semanticscholar.org/ae4d/65769b6551a51d6fc6be2f021515bffa0798.pdf):
> is actually written by my cousin. I guess I should read it from cover to cover. 😊😊

Whoa. I thought this was a joke until a saw the actual paper 😄